File: benchmarks/Csharp/Benchmarks/Rich/Scheduler.cs
#define TRACE
using System.Diagnostics;

namespace Harness.Benchmarks.Rich;

sealed class Scheduler : RBObject
{
    private TaskControlBlock taskList;
    private TaskControlBlock? currentTask;
    private int currentTaskIdentity;
    private readonly TaskControlBlock[] taskTable;

    private int queuePacketCount;
    private int holdCount;

    private int layout;

    // private const bool TRACING = false
    // Just define the TRACE symbol

    public Scheduler()
    {
        // init tracing
        layout = 0;

        // init scheduler
        queuePacketCount = 0;
        holdCount = 0;
        taskTable = new TaskControlBlock[NUM_TYPES];
        //Not needed: Arrays.Fill(taskTable, NO_TASK);
        taskList = NO_TASK;
    }


    internal void createDevice(int identity, int priority, Packet workPacket, TaskState state)
    {
        DeviceTaskDataRecord data = new DeviceTaskDataRecord();

        createTask(identity, priority, workPacket, state, (Packet workArg, RBObject wordArg) =>
        {
            DeviceTaskDataRecord dataRecord = (DeviceTaskDataRecord)wordArg;
            Packet functionWork = workArg;
            if (NO_WORK == functionWork)
            {
                if (NO_WORK == (functionWork = dataRecord.Pending))
                {
                    return markWaiting();
                }
                else
                {
                    dataRecord.Pending = NO_WORK;
                    return queuePacket(functionWork);
                }
            }
            else
            {
                dataRecord.Pending = functionWork;
                trace(functionWork.Datum);
                return holdSelf();
            }
        }, data);
    }

    internal void createHandler(int identity, int priority, Packet workPaket, TaskState state)
    {
        HandlerTaskDataRecord data = new HandlerTaskDataRecord();
        createTask(identity, priority, workPaket, state, (Packet work, RBObject word) =>
    {
        HandlerTaskDataRecord dataRecord = (HandlerTaskDataRecord)word;
        if (NO_WORK != work)
        {
            if (WORK_PACKET_KIND == work.Kind)
            {
                dataRecord.WorkInAdd(work);
            }
            else
            {
                dataRecord.DeviceInAdd(work);
            }
        }
        Packet workPacket = dataRecord.WorkIn;
        if (workPacket == NO_WORK)
        {
            return markWaiting();
        }
        else
        {
            var count = workPacket.Datum;
            if (count >= Packet.DATA_SIZE)
            {
                dataRecord.WorkIn = workPacket.Link;
                return queuePacket(workPacket);
            }
            else
            {
                Packet devicePacket = dataRecord.DeviceIn;
                if (devicePacket == NO_WORK)
                {
                    return markWaiting();
                }
                else
                {
                    dataRecord.DeviceIn = devicePacket.Link;
                    devicePacket.Datum = workPacket.Data[count];
                    workPacket.Datum = count + 1;
                    return queuePacket(devicePacket);
                }
            }
        }
    }, data);
    }

    internal void createIdler(int identity, int priority, Packet work, TaskState state)
    {
        IdleTaskDataRecord data = new IdleTaskDataRecord();
        createTask(identity, priority, work, state, (Packet workArg, RBObject wordArg) =>
        {
            var dataRecord = (IdleTaskDataRecord)wordArg;
            dataRecord.Count--;
            if (dataRecord.Count == 0)
            {
                return holdSelf();
            }
            else
            {
                if ((dataRecord.Control & 1) == 0)
                {
                    dataRecord.Control /= 2;
                    return release(DEVICE_A);
                }
                else
                {
                    dataRecord.Control = (dataRecord.Control / 2) ^ 53256;
                    return release(DEVICE_B);
                }
            }
        }, data);
    }

    internal Packet createPacket(Packet link, int identity, int kind)
    {
        return new Packet(link, identity, kind);
    }

    internal void createTask(int identity, int priority, Packet work, TaskState state, ProcessFunction aBlock, RBObject data)
    {

        TaskControlBlock t = new TaskControlBlock(taskList, identity, priority, work, state, aBlock, data);
        taskList = t;
        taskTable[identity] = t;
    }

    internal void createWorker(int identity, int priority, Packet workPaket, TaskState state)
    {
        var dataRecord = new WorkerTaskDataRecord();
        createTask(identity, priority, workPaket, state, (Packet work, RBObject word) =>
            {
                var data = (WorkerTaskDataRecord)word;
                if (NO_WORK == work)
                {
                    return markWaiting();
                }
                else
                {
                    data.Destination = (HANDLER_A == data.Destination) ? HANDLER_B : HANDLER_A;
                    work.Identity = data.Destination;
                    work.Datum = 0;
                    for (var i = 0; i < Packet.DATA_SIZE; i++)
                    {
                        data.Count++;
                        if (data.Count > 26)
                        {
                            data.Count = 1;
                        }
                        work.Data[i] = 65 + data.Count - 1;
                    }
                    return queuePacket(work);
                }
            },
        dataRecord);
    }

    public bool start()
    {
        Packet workQ;

        createIdler(IDLER, 0, NO_WORK, TaskState.createRunning());
        workQ = createPacket(NO_WORK, WORKER, WORK_PACKET_KIND);
        workQ = createPacket(workQ, WORKER, WORK_PACKET_KIND);

        createWorker(WORKER, 1000, workQ, TaskState.createWaitingWithPacket());
        workQ = createPacket(NO_WORK, DEVICE_A, DEVICE_PACKET_KIND);
        workQ = createPacket(workQ, DEVICE_A, DEVICE_PACKET_KIND);
        workQ = createPacket(workQ, DEVICE_A, DEVICE_PACKET_KIND);

        createHandler(HANDLER_A, 2000, workQ, TaskState.createWaitingWithPacket());
        workQ = createPacket(NO_WORK, DEVICE_B, DEVICE_PACKET_KIND);
        workQ = createPacket(workQ, DEVICE_B, DEVICE_PACKET_KIND);
        workQ = createPacket(workQ, DEVICE_B, DEVICE_PACKET_KIND);

        createHandler(HANDLER_B, 3000, workQ, TaskState.createWaitingWithPacket());
        createDevice(DEVICE_A, 4000, NO_WORK, TaskState.createWaiting());
        createDevice(DEVICE_B, 5000, NO_WORK, TaskState.createWaiting());

        schedule();

        return queuePacketCount == 23246 && holdCount == 9297;
    }

    internal TaskControlBlock findTask(int identity)
    {
        TaskControlBlock t = taskTable[identity];
        if (NO_TASK == t)
        {
            throw new Exception("findTask failed");
        }
        return t;
    }

    internal TaskControlBlock holdSelf()
    {
        holdCount = holdCount + 1;
        currentTask!.IsTaskHolding = true;
        return currentTask.Link;
    }

    internal TaskControlBlock queuePacket(Packet packet)
    {
        Debug.Assert(currentTask != null);
        TaskControlBlock t = findTask(packet.Identity);
        if (t == NO_TASK)
        {
            return NO_TASK;
        }

        queuePacketCount++;

        packet.Link = NO_WORK;
        packet.Identity = currentTaskIdentity;
        return t.addInputAndCheckPriority(packet, currentTask);
    }

    internal TaskControlBlock release(int identity)
    {
        Debug.Assert(currentTask != null);
        TaskControlBlock t = findTask(identity);
        if (NO_TASK == t)
        {
            return NO_TASK;
        }
        t.IsTaskHolding = false;
        if (t.Priority > currentTask.Priority)
        {
            return t;
        }
        else
        {
            return currentTask;
        }
    }

    [Conditional("TRACE")]
    internal void trace(int id)
    {
        layout--;
        if (0 >= layout)
        {
            Console.WriteLine();
            layout = 50;
        }
        Console.Write(id);

        //for (var i = 0; i < taskTable.Length; i++)
        //{
        //    TaskControlBlock? item = taskTable[i];
        //    if (item == null)
        //        Console.WriteLine($"idx {i}, no task");
        //    else
        //        Console.WriteLine($"idx {i}, {item}");
        //}
    }

    internal TaskControlBlock markWaiting()
    {
        Debug.Assert(currentTask != null);
        currentTask.IsTaskWaiting = true;
        return currentTask;
    }

    internal void schedule()
    {
        currentTask = taskList;
        while (NO_TASK != currentTask)
        {
            if (currentTask.IsTaskHoldingOrWaiting)
            {
                currentTask = currentTask.Link;
            }
            else
            {
                currentTaskIdentity = currentTask.Identity;
                trace(currentTaskIdentity);

                currentTask = currentTask.runTask();
            }
        }
    }

}

File: benchmarks/Rust/src/richards/scheduler.rs
use crate::richards::device_task_data_record::DeviceTaskDataRecord;
use crate::richards::handler_task_data_record::HandlerTaskDataRecord;
use crate::richards::idle_task_data_record::IdleTaskDataRecord;
use crate::richards::packet::{Packet, PacketBox};
use crate::richards::rb_object::RBObject;
use crate::richards::task_control_block::{TaskControlBlock, TaskControlBlockRc};
use crate::richards::task_state::TaskState;
use crate::richards::worker_task_data_record::WorkerTaskDataRecord;
use std::cell::Cell;

pub struct Scheduler {
    task_list: Option<TaskControlBlockRc>,
    current_task: Option<TaskControlBlockRc>,
    current_task_identity: usize,
    task_table: [Option<TaskControlBlockRc>; TaskControlBlock::NUM_TYPES],
    queue_packet_count: Cell<usize>,
    hold_count: Cell<usize>,
    layout: Cell<isize>,
}

impl Default for Scheduler {
    fn default() -> Self {
        Self {
            // init tracing
            layout: Cell::new(0),

            // init scheduler
            queue_packet_count: Cell::new(0),
            hold_count: Cell::new(0),
            task_table: Default::default(),

            // init task list
            current_task: None,
            current_task_identity: 0,
            task_list: None,
        }
    }
}

impl RBObject for Scheduler {}

impl Scheduler {
    const TRACING: bool = false;

    fn create_device(
        &mut self,
        identity: usize,
        priority: usize,
        work_packet: Option<PacketBox>,
        state: TaskState,
    ) {
        let mut data = DeviceTaskDataRecord::default();
        self.create_task(
            identity,
            priority,
            work_packet,
            state,
            move |sched, function_work| match function_work {
                None => match data.take_pending() {
                    None => sched.mark_waiting(),
                    Some(function_work) => {
                        data.set_pending(Packet::NO_WORK);
                        sched.queue_packet(function_work)
                    }
                },
                Some(function_work) => {
                    if Self::TRACING {
                        sched.trace(function_work.get_datum());
                    }
                    data.set_pending(Some(function_work));
                    sched.hold_self()
                }
            },
        );
    }

    fn create_handler(
        &mut self,
        identity: usize,
        priority: usize,
        work_packet: Option<PacketBox>,
        state: TaskState,
    ) {
        let mut data = HandlerTaskDataRecord::default();
        self.create_task(
            identity,
            priority,
            work_packet,
            state,
            move |sched, work| {
                if let Some(work) = work {
                    if Packet::WORK_PACKET_KIND == work.get_kind() {
                        data.work_in_add(work);
                    } else {
                        data.device_in_add(work);
                    }
                }

                match data.take_work_in() {
                    None => sched.mark_waiting(),
                    Some(mut work) => {
                        let count = work.get_datum();
                        if count >= Packet::DATA_SIZE {
                            data.set_work_in(work.take_link());
                            sched.queue_packet(work)
                        } else {
                            match data.take_device_in() {
                                None => {
                                    data.set_work_in(Some(work));
                                    sched.mark_waiting()
                                }
                                Some(mut device_packet) => {
                                    data.set_device_in(device_packet.take_link());
                                    device_packet.set_datum(work.get_data()[count]);
                                    work.set_datum(count + 1);
                                    data.set_work_in(Some(work));
                                    sched.queue_packet(device_packet)
                                }
                            }
                        }
                    }
                }
            },
        );
    }

    fn create_idler(
        &mut self,
        identity: usize,
        priority: usize,
        work: Option<PacketBox>,
        state: TaskState,
    ) {
        let mut data = IdleTaskDataRecord::default();
        self.create_task(identity, priority, work, state, move |sched, _work| {
            data.set_count(data.get_count() - 1);
            if 0 == data.get_count() {
                sched.hold_self()
            } else if 0 == data.get_control() & 1 {
                data.set_control(data.get_control() / 2);
                sched.release(Packet::DEVICE_A)
            } else {
                data.set_control((data.get_control() / 2) ^ 53256);
                sched.release(Packet::DEVICE_B)
            }
        });
    }

    fn create_packet(link: Option<PacketBox>, identity: usize, kind: usize) -> PacketBox {
        Packet::new_boxed(link, identity, kind)
    }

    fn create_task(
        &mut self,
        identity: usize,
        priority: usize,
        work: Option<PacketBox>,
        state: TaskState,
        a_block: impl 'static + FnMut(&Scheduler, Option<PacketBox>) -> Option<TaskControlBlockRc>,
    ) {
        let t = TaskControlBlock::new_rc(
            self.task_list.take(),
            identity,
            priority,
            work,
            state,
            a_block,
        );
        self.task_list = Some(t.clone());
        self.task_table[identity] = Some(t);
    }

    fn create_worker(
        &mut self,
        identity: usize,
        priority: usize,
        work_packet: Option<PacketBox>,
        state: TaskState,
    ) {
        let mut data = WorkerTaskDataRecord::default();
        self.create_task(
            identity,
            priority,
            work_packet,
            state,
            move |sched, work| match work {
                None => sched.mark_waiting(),
                Some(mut work) => {
                    data.set_destination(
                        if WorkerTaskDataRecord::HANDLER_A == data.get_destination() {
                            WorkerTaskDataRecord::HANDLER_B
                        } else {
                            WorkerTaskDataRecord::HANDLER_A
                        },
                    );
                    work.set_identity(data.get_destination());
                    work.set_datum(0);
                    for i in 0..Packet::DATA_SIZE {
                        data.set_count(data.get_count() + 1);
                        if data.get_count() > 26 {
                            data.set_count(1);
                        }
                        work.get_data_mut()[i] = 65 + data.get_count() - 1;
                    }
                    sched.queue_packet(work)
                }
            },
        );
    }

    pub fn start(&mut self) -> bool {
        self.create_idler(Self::IDLER, 0, Self::NO_WORK, TaskState::create_running());
        let mut work_q = Self::create_packet(Self::NO_WORK, Self::WORKER, Self::WORK_PACKET_KIND);
        work_q = Self::create_packet(Some(work_q), Self::WORKER, Self::WORK_PACKET_KIND);

        self.create_worker(
            Self::WORKER,
            1000,
            Some(work_q),
            TaskState::create_waiting_with_packet(),
        );
        work_q = Self::create_packet(Self::NO_WORK, Self::DEVICE_A, Self::DEVICE_PACKET_KIND);
        work_q = Self::create_packet(Some(work_q), Self::DEVICE_A, Self::DEVICE_PACKET_KIND);
        work_q = Self::create_packet(Some(work_q), Self::DEVICE_A, Self::DEVICE_PACKET_KIND);

        self.create_handler(
            Self::HANDLER_A,
            2000,
            Some(work_q),
            TaskState::create_waiting_with_packet(),
        );
        work_q = Self::create_packet(Self::NO_WORK, Self::DEVICE_B, Self::DEVICE_PACKET_KIND);
        work_q = Self::create_packet(Some(work_q), Self::DEVICE_B, Self::DEVICE_PACKET_KIND);
        work_q = Self::create_packet(Some(work_q), Self::DEVICE_B, Self::DEVICE_PACKET_KIND);

        self.create_handler(
            Self::HANDLER_B,
            3000,
            Some(work_q),
            TaskState::create_waiting_with_packet(),
        );
        self.create_device(
            Self::DEVICE_A,
            4000,
            Self::NO_WORK,
            TaskState::create_waiting(),
        );
        self.create_device(
            Self::DEVICE_B,
            5000,
            Self::NO_WORK,
            TaskState::create_waiting(),
        );

        self.schedule();

        self.queue_packet_count.get() == 23246 && self.hold_count.get() == 9297
    }

    fn find_task(&self, identity: usize) -> &TaskControlBlockRc {
        self.task_table[identity]
            .as_ref()
            .expect("find_task failed")
    }

    fn hold_self(&self) -> Option<TaskControlBlockRc> {
        self.hold_count.set(self.hold_count.get() + 1);
        if let Some(current_task) = &self.current_task {
            current_task.set_task_holding(true);
            current_task.get_link().as_ref().cloned()
        } else {
            panic!("hold_self failed")
        }
    }

    #[allow(clippy::unnecessary_wraps)]
    fn queue_packet(&self, mut packet: PacketBox) -> Option<TaskControlBlockRc> {
        self.queue_packet_count
            .set(self.queue_packet_count.get() + 1);
        let t = self.find_task(packet.get_identity());
        packet.set_link(Packet::NO_WORK);
        packet.set_identity(self.current_task_identity);
        Some(t.add_input_and_check_priority(packet, self.current_task.as_ref().unwrap(), t))
    }

    fn release(&self, identity: usize) -> Option<TaskControlBlockRc> {
        let t = self.find_task(identity);
        t.set_task_holding(false);
        if t.get_priority() > self.current_task.as_ref().unwrap().get_priority() {
            Some(t.clone())
        } else {
            self.current_task.clone()
        }
    }

    fn trace(&self, id: usize) {
        self.layout.set(self.layout.get() - 1);
        if 0 >= self.layout.get() {
            println!();
            self.layout.set(50);
        }
        println!("{id}");
    }

    fn mark_waiting(&self) -> Option<TaskControlBlockRc> {
        self.current_task.as_ref().unwrap().set_task_waiting(true);
        self.current_task.clone()
    }

    fn schedule(&mut self) {
        self.current_task = self.task_list.clone();
        while let Some(current_task) = &self.current_task {
            if current_task.is_task_holding_or_waiting() {
                self.current_task = current_task.get_link().as_ref().cloned();
            } else {
                self.current_task_identity = current_task.get_identity();
                if Self::TRACING {
                    self.trace(current_task.get_identity());
                }
                self.current_task = current_task.run_task(self);
            }
        }
    }
}
